### PR TITLE
rgw: archive zone

### PR DESCRIPTION
**Pull request by yehudasa**
Feature by @jmunhoz, rebased as original branch was based off mimic plus a few misc changes:

 - A new "archive" zone type
 - Each bucket in new zone is created as versioned
 - Synced objects are created as versioned
 - Object removals are not synced
 - Bucket are not removed but rather renamed
 - Bucket naming collision avoidance

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/rgw/rgw_data_sync.cc [hunk 1]
```diff
@@ -1706,6 +1708,20 @@ int RGWArchiveSyncModule::create_instance(CephContext *cct, const JSONFormattabl
   return 0;
 }
 
+void RGWArchiveDataSyncModule::init(RGWDataSyncEnv *sync_env, uint64_t instance_id) {
+
+  ldout(sync_env->cct, 0) << "Switching to archive zone metadata manager" << dendl;
+
+  // FIXME: this code crashes... migrating the current state (handles, md_logs,
```
Comment (jmunhoz) [inline src/rgw/rgw_data_sync.cc]: This code is legacy. We should also eliminate the support methods for metadata manager and clean it.

File: src/rgw/rgw_data_sync.h [hunk 2]
```diff
@@ -597,6 +597,7 @@ class RGWDefaultSyncModule : public RGWSyncModule {
 class RGWArchiveSyncModule : public RGWDefaultSyncModule {
 public:
   RGWArchiveSyncModule() {}
+  bool supports_writes() override { return false; }
```
Comment (jmunhoz) [inline src/rgw/rgw_data_sync.h]: It should be allowed writing in the archive zone if the user operates on the RGW(s) associated to the archive zone and it has valid permissions. It should returns true.

**Comment by jmunhoz** [comment]
@yehudasa feel free to close this dupe https://github.com/ceph/ceph/pull/24366

**Comment by jmunhoz** [comment]
Fixes: https://tracker.ceph.com/issues/36287

**Comment by yehudasa** [comment]
http://pulpito.ceph.com/yehudasa-2019-01-15_21:45:45-rgw-wip-rgw-tiering-3-distro-basic-smithi/

File: src/rgw/rgw_bucket.cc [hunk 3]
```diff
@@ -2481,6 +2481,204 @@ class RGWBucketMetadataHandler : public RGWMetadataHandler {
...
+     decode(orig_bucket, bl);
+     DECODE_FINISH(bl);
+  }
+};
+WRITE_CLASS_ENCODER(archive_meta_info)
+
+class RGWArchiveBucketMetadataHandler : public RGWBucketMetadataHandler {
+public:
+  int remove(RGWRados *store, string& entry, RGWObjVersionTracker& objv_tracker) override {
+    ldout(store->ctx(), 0) << "SKIP: bucket removal is not allowed on archive zone: bucket:" << entry << " ... proceeding to rename" << dendl;
```
Comment (theanalyst) [inline src/rgw/rgw_bucket.cc]: Increase the log level to 5 or so maybe?

File: src/rgw/rgw_data_sync.cc [hunk 4]
```diff
@@ -1686,6 +1686,83 @@ RGWCoroutine *RGWDefaultDataSyncModule::create_delete_marker(RGWDataSyncEnv *syn
...
+
+int RGWArchiveSyncModule::create_instance(CephContext *cct, const JSONFormattable& config, RGWSyncModuleInstanceRef *instance)
+{
+  instance->reset(new RGWArchiveSyncModuleInstance());
+  return 0;
+}
+
+RGWCoroutine *RGWArchiveDataSyncModule::sync_object(RGWDataSyncEnv *sync_env, RGWBucketInfo& bucket_info, rgw_obj_key& key, std::optional<uint64_t> versioned_epoch, rgw_zone_set *zones_trace)
+{
+  ldout(sync_env->cct, 0) << "SYNC_ARCHIVE: sync_object: b=" << bucket_info.bucket << " k=" << key << " versioned_epoch=" << versioned_epoch.value_or(0) << dendl;
```
Comment (theanalyst) [inline src/rgw/rgw_data_sync.cc]: higher log level?

Comment (yehudasa) [reply src/rgw/rgw_data_sync.cc]: @theanalyst fixed

File: src/rgw/rgw_bucket.cc [hunk 5]
```diff
@@ -2481,6 +2481,204 @@ class RGWBucketMetadataHandler : public RGWMetadataHandler {
...
+
+    /* clean up old stuff */
+
+    ret = rgw_unlink_bucket(store, be.owner, tenant_name, bucket_name, false);
+    if (ret < 0) {
+        lderr(store->ctx()) << "could not unlink bucket=" << entry << " owner=" << be.owner << dendl;
+    }
+
+    // if (ret == -ECANCELED) it means that there was a race here, and someone
+    // wrote to the bucket entrypoint just before we removed it. The question is
```
Comment (theanalyst) [inline src/rgw/rgw_bucket.cc]: We don't seem to be specially handling -ECANCELED as mentioned in the comment is that intentional?

File: src/rgw/rgw_bucket.cc [hunk 6]
```diff
@@ -2481,6 +2481,204 @@ class RGWBucketMetadataHandler : public RGWMetadataHandler {
...
+    ret = rgw_delete_system_obj(store, store->svc.zone->get_zone_params().domain_root, RGW_BUCKET_INSTANCE_MD_PREFIX + meta_name, NULL);
+
+    /* idempotent */
+
+    return 0;
+  }
+
+  int put(RGWRados *store, string& entry, RGWObjVersionTracker& objv_tracker,
+          real_time mtime, JSONObj *obj, sync_type_t sync_type) override {
+    if (entry.find("-deleted-") != string::npos) {
```
Comment (theanalyst) [inline src/rgw/rgw_bucket.cc]: are there cases where a normal bucket name containing -deleted- would actually trigger this path in the archive zone

File: src/rgw/rgw_bucket.cc (hunk 5 above)
Comment (yehudasa) [reply src/rgw/rgw_bucket.cc]: Right. However, these can't really happen concurrently as they go through the meta sync path (in archive zone), so it's a non issue.

File: src/rgw/rgw_bucket.cc (hunk 6 above)
Comment (yehudasa) [reply src/rgw/rgw_bucket.cc]: in theory yes, but will behave the same. A normal bucket at this point will already have had `remove()` called on it. Note that it doesn't remove it completely, just renames it, and the code here deals with how to do the renaming correctly.

**Comment by yehudasa** [comment]
http://pulpito.ceph.com/yehudasa-2019-01-25_16:08:06-rgw-wip-archive-zone-3-distro-basic-smithi/

**Pull request merged**